roachtest: decommissioning test

tbg opened the pull request:
Add a test that decommissions and wipes the first node in a cluster
repeatedly, alternating between the node being shut down gracefully
before and after the decommissioning operation, while some light load
is running against the cluster (to manually verify that the qps don't
dip too much).

a TODO is to verify that the logs don't contain the messages that would
spam the log before https://github.com/cockroachdb/cockroach/pull/23605.
I wonder if we should really start grepping the logs. An alternative is
to introduce a metric that would have signaled this and check that
instead.

another TODO is to actually make this work against remote clusters; it's
currently blocked on not being able to look up the internal IP. I think
we need to expose the `expander` functionality in `roachprod` to avoid
having to plug these random holes in every PR, though I want
@petermattis to comment on that suggestion first. (Note that I could
also manually execute `roachprod pgurl` and then parse the address
out, but that's just beating around the bush).

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/23769)
<!-- Reviewable:end -->

petermattis commented:
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/cmd/roachtest/decommission.go, line 140 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23769#-L7Rxkhdlu4RY0c7QuJn:-L7Rxkhdlu4RY0c7QuJo:b-4h9co9) ([raw file](https://github.com/cockroachdb/cockroach/blob/369eaf8f923751514c63257e37a3c9febcffd6b6/pkg/cmd/roachtest/decommission.go#L140)):*
> ```Go
> 			}
> 			// TODO(tschottdorf): need to add a roachprod expander to make this work without the hack.
> 			// It would be a lot easier if all that stuff weren't hidden away in roachprod.
> ```

So you want something like `{ipaddr:%[1]d}:{pgport:%[1]d}`? Easy enough. The biggest irritation is that this is in a different repo. Do you have a suggestion for exposing the expander stuff in a way that is easier to extend?

---

*[pkg/cmd/roachtest/decommission.go, line 162 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23769#-L7RyhCh2-RG5nPiX3zM:-L7RyhCh2-RG5nPiX3zN:befz54y) ([raw file](https://github.com/cockroachdb/cockroach/blob/369eaf8f923751514c63257e37a3c9febcffd6b6/pkg/cmd/roachtest/decommission.go#L162)):*
> ```Go
> func init() {
> 	const nodes = 4
> 	duration := 5 * time.Hour
> ```

That's a lot of decommissioning! 

---

*[pkg/cmd/roachtest/decommission.go, line 165 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23769#-L7RzVxPrYE0fPBlgxrB:-L7RzVxPrYE0fPBlgxrC:bfhets2) ([raw file](https://github.com/cockroachdb/cockroach/blob/369eaf8f923751514c63257e37a3c9febcffd6b6/pkg/cmd/roachtest/decommission.go#L165)):*
> ```Go
> 
> 	tests.Add(
> 		fmt.Sprintf("decommission/nodes=%d/duration=%s", nodes, duration),
> ```

We haven't been adding duration to the other test names. Should we? I like consistency. Perhaps the duration should be part of the `testSpec` so we can optionally print it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23769)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/cmd/roachtest/decommission.go, line 140 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23769#-L7Rxkhdlu4RY0c7QuJn:-L7UhuuX5g0Xq1w32oTD:b-ahkz73) ([raw file](https://github.com/cockroachdb/cockroach/blob/369eaf8f923751514c63257e37a3c9febcffd6b6/pkg/cmd/roachtest/decommission.go#L140)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

So you want something like `{ipaddr:%[1]d}:{pgport:%[1]d}`? Easy enough. The biggest irritation is that this is in a different repo. Do you have a suggestion for exposing the expander stuff in a way that is easier to extend?
</blockquote></details>

I was thinking to introduce `roachprod expand <selector> <string>` that just calls into the expander (along with a pass-through in `*cluster`), so all we'll ever have to do after that is add new expanders. `roachprod pgurl` etc would just be sugar around `roachprod expand`.

---

*[pkg/cmd/roachtest/decommission.go, line 162 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23769#-L7RyhCh2-RG5nPiX3zM:-L7Uip41DWvmp4_12ctH:bzhp6sw) ([raw file](https://github.com/cockroachdb/cockroach/blob/369eaf8f923751514c63257e37a3c9febcffd6b6/pkg/cmd/roachtest/decommission.go#L162)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

That's a lot of decommissioning! 
</blockquote></details>

Gotta make it count! Note that a decommissioning step may take a long time if the data set on that node is large.

---

*[pkg/cmd/roachtest/decommission.go, line 165 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23769#-L7RzVxPrYE0fPBlgxrB:-L7Ui1O4181mcWC1xgPl:b-jgrjvk) ([raw file](https://github.com/cockroachdb/cockroach/blob/369eaf8f923751514c63257e37a3c9febcffd6b6/pkg/cmd/roachtest/decommission.go#L165)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

We haven't been adding duration to the other test names. Should we? I like consistency. Perhaps the duration should be part of the `testSpec` so we can optionally print it.
</blockquote></details>

I'm hesitant to bake these things in, as a duration makes sense for this test but little for others (one of our warts in the acceptance tests). I too like consistency, but would hold off on casting a mold just yet. Maybe instead of an explicit duration the testSpec can provide a time budget and the test must terminate within the budget, or fail when it doesn't manage to. This makes sense for both time-based and action-based tests: run this restore in less than four hours to pass, run this decommissioning test for four hours. Either way, I'd leave all of that for another PR.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23769)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/cmd/roachtest/decommission.go, line 140 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23769#-L7Rxkhdlu4RY0c7QuJn:-L7Uk3sA43obVL_VQNRS:b-nh6n5r) ([raw file](https://github.com/cockroachdb/cockroach/blob/369eaf8f923751514c63257e37a3c9febcffd6b6/pkg/cmd/roachtest/decommission.go#L140)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

I was thinking to introduce `roachprod expand <selector> <string>` that just calls into the expander (along with a pass-through in `*cluster`), so all we'll ever have to do after that is add new expanders. `roachprod pgurl` etc would just be sugar around `roachprod expand`.
</blockquote></details>

What is `<selector>` in this example? I'm probably missing something basic about your proposal. Adding the new expansion to `roachprod/install/expander.go` causes it to automatically become available to both `roachprod ssh` and `roachprod start`. I suppose we can add a `roachprod expand` as well, though you wouldn't need it below, right?

---

*[pkg/cmd/roachtest/decommission.go, line 162 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23769#-L7RyhCh2-RG5nPiX3zM:-L7UlHTn5odRSAMPGmaW:bxopnr9) ([raw file](https://github.com/cockroachdb/cockroach/blob/369eaf8f923751514c63257e37a3c9febcffd6b6/pkg/cmd/roachtest/decommission.go#L162)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Gotta make it count! Note that a decommissioning step may take a long time if the data set on that node is large.
</blockquote></details>

How many decommissionings does the test get through in 5h?

---

*[pkg/cmd/roachtest/decommission.go, line 165 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23769#-L7RzVxPrYE0fPBlgxrB:-L7Ul3Nx3cwnvFRIQ3T4:baplq09) ([raw file](https://github.com/cockroachdb/cockroach/blob/369eaf8f923751514c63257e37a3c9febcffd6b6/pkg/cmd/roachtest/decommission.go#L165)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

I'm hesitant to bake these things in, as a duration makes sense for this test but little for others (one of our warts in the acceptance tests). I too like consistency, but would hold off on casting a mold just yet. Maybe instead of an explicit duration the testSpec can provide a time budget and the test must terminate within the budget, or fail when it doesn't manage to. This makes sense for both time-based and action-based tests: run this restore in less than four hours to pass, run this decommissioning test for four hours. Either way, I'd leave all of that for another PR.
</blockquote></details>

Ack. We also have some tests which specify `nodes=%d` while others do not. Fine to leave as-is for this PR.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23769#-:-L7UlP0g2i5ujMDDieRm:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/cmd/roachtest/decommission.go, line 140 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23769#-L7Rxkhdlu4RY0c7QuJn:-L7V2AuREpBhSAFEW7Kv:bwrxfpi) ([raw file](https://github.com/cockroachdb/cockroach/blob/369eaf8f923751514c63257e37a3c9febcffd6b6/pkg/cmd/roachtest/decommission.go#L140)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

What is `<selector>` in this example? I'm probably missing something basic about your proposal. Adding the new expansion to `roachprod/install/expander.go` causes it to automatically become available to both `roachprod ssh` and `roachprod start`. I suppose we can add a `roachprod expand` as well, though you wouldn't need it below, right?
</blockquote></details>

Yeah, `roachprod expand` isn't necessary here. I decided to just introduce `c.PGUrl` and parse what I need from that.

---

*[pkg/cmd/roachtest/decommission.go, line 162 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23769#-L7RyhCh2-RG5nPiX3zM:-L7ePagam0K9Z3k-3WW2:b-7bl9yk) ([raw file](https://github.com/cockroachdb/cockroach/blob/369eaf8f923751514c63257e37a3c9febcffd6b6/pkg/cmd/roachtest/decommission.go#L162)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

How many decommissionings does the test get through in 5h?
</blockquote></details>

Not sure, didn't really have it in me to run this for 5h through my laptop which would definitely go to standby at some point. Reducing to 1h. The cycles are pretty fast but note the comment that I added (includes some mild rage, apologies) on how we don't manage to get data on the nodes that are being cycled. Definitely still work left to do on this test but I'll merge it for now.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23769)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/cmd/roachtest/decommission.go, line 167 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23769#-L7er9mD2roMJedEuwbO:-L7er9mD2roMJedEuwbP:b-y22wxl) ([raw file](https://github.com/cockroachdb/cockroach/blob/f68b34aa1eb12eee17b8f9123cef5745ceba97fd/pkg/cmd/roachtest/decommission.go#L167)):*
> ```Go
> 			//
> 			// TODO(petermattis): allow overriding args for the start invocation
> 			// or provide the required functionality otherwise.
> ```

Can you expand on which args you want to override? You're override the `--join` flag above? Overriding the `--store` flag seems feasible given the `{store-dir}` expansion. Alternately, there is the `--racks` flag to start which can assign nodes in the cluster to "racks" (node `i` is assigned to rack `i % numRacks` via a `racks=` locality tag).

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23769)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
TFTR!

---

Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/cmd/roachtest/decommission.go, line 167 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23769#-L7er9mD2roMJedEuwbO:-L7euSJSBzZ6S4-VlNNk:bkbuup3) ([raw file](https://github.com/cockroachdb/cockroach/blob/f68b34aa1eb12eee17b8f9123cef5745ceba97fd/pkg/cmd/roachtest/decommission.go#L167)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Can you expand on which args you want to override? You're override the `--join` flag above? Overriding the `--store` flag seems feasible given the `{store-dir}` expansion. Alternately, there is the `--racks` flag to start which can assign nodes in the cluster to "racks" (node `i` is assigned to rack `i % numRacks` via a `racks=` locality tag).
</blockquote></details>

I don't even actually care that much specifically. I just want to achieve that the first node (which is the one I'm repeatedly decommissioning, wiping and restarting) is preferred for replica placement whenever it is around (but not required, or you can't decommission it). I'm not sure how I would override the store-dir since I can't specify the arg twice (I tried passing in `--locality` but you can't pass it twice). I could pass a second store and make that one preferred, but that seemed one hack to many. I'm not sure how `racks` helps. It looks to have the same problems as `--locality` namely that I'm not in charge of which node gets what. I would have to pick the nodes so that they have some desired combination of racks, but I don't want to have to do that. We'll need this type of control in the future.

I don't want to figure this out in this PR, let's revisit this separately.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23769#-:-L7fGB8XE1d_ytpF5kLL:b-814g8k)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/cmd/roachtest/decommission.go, line 167 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23769#-L7er9mD2roMJedEuwbO:-L7fGfNnEVwRhDTnKoQA:b-sb0kxk) ([raw file](https://github.com/cockroachdb/cockroach/blob/f68b34aa1eb12eee17b8f9123cef5745ceba97fd/pkg/cmd/roachtest/decommission.go#L167)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

I don't even actually care that much specifically. I just want to achieve that the first node (which is the one I'm repeatedly decommissioning, wiping and restarting) is preferred for replica placement whenever it is around (but not required, or you can't decommission it). I'm not sure how I would override the store-dir since I can't specify the arg twice (I tried passing in `--locality` but you can't pass it twice). I could pass a second store and make that one preferred, but that seemed one hack to many. I'm not sure how `racks` helps. It looks to have the same problems as `--locality` namely that I'm not in charge of which node gets what. I would have to pick the nodes so that they have some desired combination of racks, but I don't want to have to do that. We'll need this type of control in the future.

I don't want to figure this out in this PR, let's revisit this separately.
</blockquote></details>

Agreed, no need to figure it out in this PR. Can you file an issue? I'm still not understanding what you're trying to do. You're right that passing the `--store` flag twice doesn't do the right thing here as it creates a second store.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23769)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/cmd/roachtest/decommission.go, line 167 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23769#-L7er9mD2roMJedEuwbO:-L7fXzVkBgLz2s-qCN3t:b-tb20hg) ([raw file](https://github.com/cockroachdb/cockroach/blob/f68b34aa1eb12eee17b8f9123cef5745ceba97fd/pkg/cmd/roachtest/decommission.go#L167)):*
Just that:

> I just want to achieve that the first node (which is the one I'm repeatedly decommissioning, wiping and restarting) is preferred for replica placement whenever it is around

For example, I want four nodes and only the first one has a locality set. That way, the allocator will always try to move replicas to that node, but if the node gets decommissioned, it will move them away from it (hopefully).

I'm not sure which issue to file. I think not being able to control what gets passed to `cockroach start` is bad. But I'm not convinced that needs to be fixed to address that specific issue (though that's the most straightforward way I can imagine).

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23769)*
<!-- Sent from Reviewable.io -->

tbg committed:
roachtest: decommissioning test

Add a test that decommissions and wipes the first node in a cluster
repeatedly, alternating between the node being shut down gracefully
before and after the decommissioning operation, while some light load
is running against the cluster (to manually verify that the qps don't
dip too much).

There are some TODOs inside that would make this test a lot more useful.

Release note: None

petermattis commented:
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/cmd/roachtest/decommission.go, line 167 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23769#-L7er9mD2roMJedEuwbO:-L7fZ4ZC4tERIZKYqfiE:b9cgijo) ([raw file](https://github.com/cockroachdb/cockroach/blob/f68b34aa1eb12eee17b8f9123cef5745ceba97fd/pkg/cmd/roachtest/decommission.go#L167)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Just that:

> I just want to achieve that the first node (which is the one I'm repeatedly decommissioning, wiping and restarting) is preferred for replica placement whenever it is around

For example, I want four nodes and only the first one has a locality set. That way, the allocator will always try to move replicas to that node, but if the node gets decommissioned, it will move them away from it (hopefully).

I'm not sure which issue to file. I think not being able to control what gets passed to `cockroach start` is bad. But I'm not convinced that needs to be fixed to address that specific issue (though that's the most straightforward way I can imagine).
</blockquote></details>

I'm fine with adding control over the flags passed to `start`. 

To solve the immediate issue, I think you could do something like `roachprod start <cluster> --racks=4` and then specify a zone config which required a replica on `+racks=0`.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23769)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/cmd/roachtest/decommission.go, line 167 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23769#-L7er9mD2roMJedEuwbO:-L7feGUF5VF8TjSq0_7e:b3j1g4e) ([raw file](https://github.com/cockroachdb/cockroach/blob/f68b34aa1eb12eee17b8f9123cef5745ceba97fd/pkg/cmd/roachtest/decommission.go#L167)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I'm fine with adding control over the flags passed to `start`. 

To solve the immediate issue, I think you could do something like `roachprod start <cluster> --racks=4` and then specify a zone config which required a replica on `+racks=0`.
</blockquote></details>

But wouldn't that *require* one at rack zero? I.e. when decommissioning, it wouldn't place that replica elsewhere and so it would hang forever.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23769)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
cc @garvitjuniwal @neeral

petermattis commented:
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/cmd/roachtest/decommission.go, line 167 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23769#-L7er9mD2roMJedEuwbO:-L7iWofy6fOunO2RHCH-:b-koekeg) ([raw file](https://github.com/cockroachdb/cockroach/blob/f68b34aa1eb12eee17b8f9123cef5745ceba97fd/pkg/cmd/roachtest/decommission.go#L167)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

But wouldn't that *require* one at rack zero? I.e. when decommissioning, it wouldn't place that replica elsewhere and so it would hang forever.
</blockquote></details>

I can't recall for sure. You're probably right.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23769)*
<!-- Sent from Reviewable.io -->
